File: scripts/dev/pds_cli.py
#!/usr/bin/env python3
"""PDS dev helper: thin wrapper around the native kaszlak binary plus a few HTTP flows.

Environment:
  NO_COLOR                 Disable ANSI colors (also implied when stdout is not a TTY).
  PDS_CLI_JSON=1           JSON status lines instead of ANSI formatting.
  PDS_CREATE_ACCOUNT_PASSWORD / PDS_POST_PASSWORD — optional alternatives to ``--password``.
  PDS_URL, PDS_DATA_DIR, PDS_BIN — see defaults below.

Leading ``--json`` / ``-j`` / ``--no-color`` are stripped before argparse so invocations like
``pds_cli.py --json account create ...`` work. The kaszlak binary requires ``kaszlak <command> [flags]``;
this script builds argv in that order.
"""
from __future__ import annotations

import argparse
import datetime
import json
import os
import subprocess
import sys

import requests

# --- Configuration ---
PDS_URL = os.environ.get("PDS_URL", "http://localhost:2583")
REPO_ROOT = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
DATA_DIR = os.environ.get("PDS_DATA_DIR", os.path.join(REPO_ROOT, "data"))
BIN_PATH = os.environ.get("PDS_BIN", os.path.join(REPO_ROOT, "build/bin/kaszlak"))

PDS_CLI_JSON = os.environ.get("PDS_CLI_JSON", "").lower() in ("1", "true", "yes")
PDS_CLI_NO_COLOR = False
USE_COLOR = False


def _strip_wrapper_prefix_flags() -> None:
    """Remove leading --json / -j / --no-color from sys.argv (in place)."""
    global PDS_CLI_JSON, PDS_CLI_NO_COLOR
    i = 1
    while i < len(sys.argv):
        a = sys.argv[i]
        if a in ("--json", "-j"):
            PDS_CLI_JSON = True
            sys.argv.pop(i)
            continue
        if a == "--no-color":
            PDS_CLI_NO_COLOR = True
            sys.argv.pop(i)
            continue
        if a.startswith("-"):
            break
        break


def _configure_output() -> None:
    global USE_COLOR
    if os.environ.get("NO_COLOR"):
        USE_COLOR = False
    elif PDS_CLI_NO_COLOR:
        USE_COLOR = False
    elif PDS_CLI_JSON:
        USE_COLOR = False
    else:
        USE_COLOR = sys.stdout.isatty()


class Colors:
    HEADER = "\033[95m"
    OKBLUE = "\033[94m"
    OKCYAN = "\033[96m"
    OKGREEN = "\033[92m"
    WARNING = "\033[93m"
    FAIL = "\033[91m"
    ENDC = "\033[0m"
    BOLD = "\033[1m"
    UNDERLINE = "\033[4m"


def print_success(msg: str) -> None:
    if PDS_CLI_JSON:
        print(json.dumps({"kind": "ok", "message": msg}))
        return
    if USE_COLOR:
        print(f"{Colors.OKGREEN}OK: {msg}{Colors.ENDC}")
    else:
        print(f"OK: {msg}")


def print_info(msg: str) -> None:
    if PDS_CLI_JSON:
        print(json.dumps({"kind": "info", "message": msg}))
        return
    if USE_COLOR:
        print(f"{Colors.OKBLUE}{msg}{Colors.ENDC}")
    else:
        print(msg)


def print_error(msg: str, **extra: object) -> None:
    if PDS_CLI_JSON:
        payload: dict[str, object] = {"kind": "error", "message": msg}
        for k, v in extra.items():
            if v is not None and v != "":
                payload[k] = v
        print(json.dumps(payload))
        return
    if USE_COLOR:
        print(f"{Colors.FAIL}Error: {msg}{Colors.ENDC}", file=sys.stderr)
    else:
        print(f"Error: {msg}", file=sys.stderr)


def run_kaszlak_cli(args: list[str]) -> int:
    """Run kaszlak; ``args`` must start with the subcommand, e.g. ``['account', 'create', ...]``."""
    if not args:
        print_error("Internal error: empty kaszlak argv")
        return 1
    if not os.path.exists(BIN_PATH):
        print_error(f"Binary not found at: {BIN_PATH}")
        print_info("Build the project first or set PDS_BIN.")
        return 1

    subcommand = args[0]
    tail = args[1:]
    cmd = [
        BIN_PATH,
        subcommand,
        "--verbose",
        "--data-dir",
        DATA_DIR,
        "--config",
        "/tmp/missing_cli_config.json",
    ] + tail

    try:
        result = subprocess.run(cmd, capture_output=True, text=True)
        if result.returncode != 0:
            print_error(
                "kaszlak command failed",
                command=" ".join(cmd),
                stderr=(result.stderr or "").strip(),
                exit_code=result.returncode,
            )
            if not PDS_CLI_JSON and result.stdout:
                sys.stdout.write(result.stdout)
            return result.returncode
        if result.stdout:
            sys.stdout.write(result.stdout)
        return 0
    except OSError as e:
        print_error(f"Failed to run binary: {e}")
        return 1


def login(handle: str, password: str):
    print_info(f"Logging in as {handle}...")
    try:
        resp = requests.post(
            f"{PDS_URL}/xrpc/com.atproto.server.createSession",
            json={"identifier": handle, "password": password},
            timeout=10,
        )
        if resp.status_code != 200:
            print_error(f"Login failed: {resp.text}", status_code=resp.status_code)
            return None
        return resp.json()
    except requests.exceptions.ConnectionError:
        print_error(f"Could not connect to PDS at {PDS_URL}")
        return None


def create_record(session: dict, collection: str, record: dict):
    try:
        resp = requests.post(
            f"{PDS_URL}/xrpc/com.atproto.repo.createRecord",
            headers={"Authorization": f"Bearer {session['accessJwt']}"},
            json={"repo": session["did"], "collection": collection, "record": record},
            timeout=10,
        )
        if resp.status_code != 200:
            print_error(f"Failed to create record: {resp.text}", status_code=resp.status_code)
            return None
        return resp.json()
    except OSError as e:
        print_error(f"Exception during record creation: {e}")
        return None


def handle_account_create(args: argparse.Namespace) -> int:
    print_info(f"Creating account for {args.handle}...")
    password = args.password or os.environ.get("PDS_CREATE_ACCOUNT_PASSWORD")
    if not password:
        print_error(
            "Missing password: use --password or set PDS_CREATE_ACCOUNT_PASSWORD in the environment."
        )
        return 2
    rc = run_kaszlak_cli(
        [
            "account",
            "create",
            "--email",
            args.email,
            "--handle",
            args.handle,
            "--password",
            password,
        ]
    )
    if rc == 0:
        print_success(f"Account {args.handle} created successfully.")
    else:
        print_error("Failed to create account.", exit_code=rc)
    return rc


def handle_post_create(args: argparse.Namespace) -> int:
    password = args.password or os.environ.get("PDS_POST_PASSWORD")
    if not password:
        print_error(
            "Missing password: use --password or set PDS_POST_PASSWORD in the environment."
        )
        return 2
    session = login(args.handle, password)
    if not session:
        return 1

    print_info("Creating post...")
    now = datetime.datetime.now(datetime.timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")
    record = {"$type": "app.bsky.feed.post", "text": args.text, "createdAt": now}

    res = create_record(session, "app.bsky.feed.post", record)
    if res:
        print_success(f"Post created! URI: {res.get('uri')}")
        if PDS_CLI_JSON:
            print(json.dumps({"kind": "result", "uri": res.get("uri"), "cid": res.get("cid")}))
        else:
            print(f"CID: {res.get('cid')}")
        return 0
    return 1


def handle_profile_update(args: argparse.Namespace) -> int:
    password = args.password or os.environ.get("PDS_POST_PASSWORD")
    if not password:
        print_error(
            "Missing password: use --password or set PDS_POST_PASSWORD in the environment."
        )
        return 2
    session = login(args.handle, password)
    if not session:
        return 1

    print_info("Updating profile...")
    record = {
        "$type": "app.bsky.actor.profile",
        "displayName": args.name,
        "description": args.description,
    }

    res = create_record(session, "app.bsky.actor.profile", record)
    if res:
        print_success("Profile updated!")
        return 0
    return 1


def main() -> int:
    _strip_wrapper_prefix_flags()
    _configure_output()

    parser = argparse.ArgumentParser(
        description="PDS CLI Tool — dev wrapper around kaszlak and local XRPC.",
        epilog="Machine output: PDS_CLI_JSON=1 or leading --json. "
        "Passwords: --password or PDS_CREATE_ACCOUNT_PASSWORD / PDS_POST_PASSWORD.",
    )
    subparsers = parser.add_subparsers(dest="command", help="Available commands")

    parser_account = subparsers.add_parser("account", help="Account management")
    account_subparsers = parser_account.add_subparsers(dest="subcommand", help="Account actions")

    create_account_parser = account_subparsers.add_parser("create", help="Create a new account")
    create_account_parser.add_argument("handle", help="User handle (e.g. alice.test)")
    create_account_parser.add_argument("email", help="User email")
    create_account_parser.add_argument(
        "--password",
        default=None,
        help="Account password (default: PDS_CREATE_ACCOUNT_PASSWORD env)",
    )

    parser_post = subparsers.add_parser("post", help="Post management")
    post_subparsers = parser_post.add_subparsers(dest="subcommand", help="Post actions")

    create_post_parser = post_subparsers.add_parser("create", help="Create a text post")
    create_post_parser.add_argument("handle", help="User handle")
    create_post_parser.add_argument("text", help="Post content")
    create_post_parser.add_argument(
        "--password",
        default=None,
        help="Session password (default: PDS_POST_PASSWORD env)",
    )

    parser_profile = subparsers.add_parser("profile", help="Profile management")
    profile_subparsers = parser_profile.add_subparsers(dest="subcommand", help="Profile actions")

    update_profile_parser = profile_subparsers.add_parser("update", help="Update profile")
    update_profile_parser.add_argument("handle", help="User handle")
    update_profile_parser.add_argument("--name", required=True, help="Display Name")
    update_profile_parser.add_argument("--description", required=True, help="Description")
    update_profile_parser.add_argument(
        "--password",
        default=None,
        help="Session password (default: PDS_POST_PASSWORD env)",
    )

    args = parser.parse_args()

    if args.command == "account" and args.subcommand == "create":
        return handle_account_create(args)
    if args.command == "post" and args.subcommand == "create":
        return handle_post_create(args)
    if args.command == "profile" and args.subcommand == "update":
        return handle_profile_update(args)

    parser.print_help()
    return 2


if __name__ == "__main__":
    sys.exit(main())

File: scripts/dev/pds_cli.ts
#!/usr/bin/env -S deno run -A

let jsonOutput = ["1", "true", "yes"].includes((Deno.env.get("PDS_CLI_JSON") || "").toLowerCase());
let noColor = false;
let useColor = false;

const pdsUrl = Deno.env.get("PDS_URL") || "http://localhost:2583";
const repoRoot = await (async () => {
  const output = await new Deno.Command("git", {
    args: ["rev-parse", "--show-toplevel"],
    stdout: "piped",
    stderr: "null",
  }).output();
  return new TextDecoder().decode(output.stdout).trim() ||
    new URL("../..", import.meta.url).pathname;
})();
const dataDir = Deno.env.get("PDS_DATA_DIR") || `${repoRoot}/data`;
const binPath = Deno.env.get("PDS_BIN") || `${repoRoot}/build/bin/kaszlak`;

function stripWrapperFlags(argv: string[]): string[] {
  const copy = [...argv];
  while (copy.length > 0) {
    if (copy[0] === "--json" || copy[0] === "-j") {
      jsonOutput = true;
      copy.shift();
    } else if (copy[0] === "--no-color") {
      noColor = true;
      copy.shift();
    } else {
      break;
    }
  }
  return copy;
}

function configureOutput() {
  useColor = !Deno.env.get("NO_COLOR") && !noColor && !jsonOutput && Deno.stdout.isTerminal();
}

const colors = {
  green: "\x1b[92m",
  blue: "\x1b[94m",
  red: "\x1b[91m",
  reset: "\x1b[0m",
};

function printJson(kind: string, message: string, extra: Record<string, unknown> = {}) {
  console.log(JSON.stringify({ kind, message, ...extra }));
}

function printSuccess(message: string) {
  if (jsonOutput) return printJson("ok", message);
  console.log(useColor ? `${colors.green}OK: ${message}${colors.reset}` : `OK: ${message}`);
}

function printInfo(message: string) {
  if (jsonOutput) return printJson("info", message);
  console.log(useColor ? `${colors.blue}${message}${colors.reset}` : message);
}

function printError(message: string, extra: Record<string, unknown> = {}) {
  if (jsonOutput) return printJson("error", message, extra);
  console.error(useColor ? `${colors.red}Error: ${message}${colors.reset}` : `Error: ${message}`);
}

function takeValue(argv: string[], index: number, flag: string): string {
  const value = argv[index + 1];
  if (!value) {
    printError(`${flag} requires a value`);
    Deno.exit(2);
  }
  return value;
}

async function runKaszlak(args: string[]): Promise<number> {
  if (args.length === 0) {
    printError("Internal error: empty kaszlak argv");
    return 1;
  }
  try {
    await Deno.stat(binPath);
  } catch {
    printError(`Binary not found at: ${binPath}`);
    printInfo("Build the project first or set PDS_BIN.");
    return 1;
  }

  const [subcommand, ...tail] = args;
  const cmd = [
    subcommand,
    "--verbose",
    "--data-dir",
    dataDir,
    "--config",
    "/tmp/missing_cli_config.json",
    ...tail,
  ];
  const output = await new Deno.Command(binPath, {
    args: cmd,
    stdout: "piped",
    stderr: "piped",
  }).output();
  const stdout = new TextDecoder().decode(output.stdout);
  const stderr = new TextDecoder().decode(output.stderr);
  if (output.code !== 0) {
    printError("kaszlak command failed", {
      command: [binPath, ...cmd].join(" "),
      stderr: stderr.trim(),
      exit_code: output.code,
    });
    if (!jsonOutput && stdout) console.log(stdout);
    return output.code;
  }
  if (stdout) console.log(stdout.trimEnd());
  return 0;
}

async function xrpcPost(method: string, body: Record<string, unknown>, token?: string) {
  const response = await fetch(`${pdsUrl}/xrpc/${method}`, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      ...(token ? { Authorization: `Bearer ${token}` } : {}),
    },
    body: JSON.stringify(body),
  });
  const text = await response.text();
  const payload = text ? JSON.parse(text) : {};
  if (!response.ok) {
    printError(`XRPC ${method} failed: ${text}`, { status_code: response.status });
    return null;
  }
  return payload;
}

async function login(handle: string, password: string) {
  printInfo(`Logging in as ${handle}...`);
  try {
    return await xrpcPost("com.atproto.server.createSession", { identifier: handle, password });
  } catch {
    printError(`Could not connect to PDS at ${pdsUrl}`);
    return null;
  }
}

async function createRecord(
  session: Record<string, string>,
  collection: string,
  record: Record<string, unknown>,
) {
  return await xrpcPost("com.atproto.repo.createRecord", {
    repo: session.did,
    collection,
    record,
  }, session.accessJwt);
}

async function handleAccountCreate(argv: string[]): Promise<number> {
  const handle = argv[0];
  const email = argv[1];
  let password = Deno.env.get("PDS_CREATE_ACCOUNT_PASSWORD") || "";
  for (let i = 2; i < argv.length; i++) {
    if (argv[i] === "--password") password = takeValue(argv, i++, argv[i]);
    else {
      printError(`Unexpected argument: ${argv[i]}`);
      return 2;
    }
  }
  if (!handle || !email) {
    printError("Usage: account create HANDLE EMAIL [--password PASSWORD]");
    return 2;
  }
  if (!password) {
    printError("Missing password: use --password or set PDS_CREATE_ACCOUNT_PASSWORD.");
    return 2;
  }
  printInfo(`Creating account for ${handle}...`);
  const rc = await runKaszlak([
    "account",
    "create",
    "--email",
    email,
    "--handle",
    handle,
    "--password",
    password,
  ]);
  if (rc === 0) printSuccess(`Account ${handle} created successfully.`);
  else printError("Failed to create account.", { exit_code: rc });
  return rc;
}

async function handlePostCreate(argv: string[]): Promise<number> {
  const handle = argv[0];
  const text = argv[1];
  let password = Deno.env.get("PDS_POST_PASSWORD") || "";
  for (let i = 2; i < argv.length; i++) {
    if (argv[i] === "--password") password = takeValue(argv, i++, argv[i]);
    else {
      printError(`Unexpected argument: ${argv[i]}`);
      return 2;
    }
  }
  if (!handle || !text) {
    printError("Usage: post create HANDLE TEXT [--password PASSWORD]");
    return 2;
  }
  if (!password) {
    printError("Missing password: use --password or set PDS_POST_PASSWORD.");
    return 2;
  }
  const session = await login(handle, password);
  if (!session) return 1;
  const result = await createRecord(session, "app.bsky.feed.post", {
    "$type": "app.bsky.feed.post",
    text,
    createdAt: new Date().toISOString().replace(/\.\d{3}Z$/, "Z"),
  });
  if (!result) return 1;
  printSuccess(`Post created! URI: ${result.uri}`);
  if (jsonOutput) printJson("result", "post", { uri: result.uri, cid: result.cid });
  else console.log(`CID: ${result.cid}`);
  return 0;
}

async function handleProfileUpdate(argv: string[]): Promise<number> {
  const handle = argv[0];
  let name = "";
  let description = "";
  let password = Deno.env.get("PDS_POST_PASSWORD") || "";
  for (let i = 1; i < argv.length; i++) {
    if (argv[i] === "--name") name = takeValue(argv, i++, argv[i]);
    else if (argv[i] === "--description") description = takeValue(argv, i++, argv[i]);
    else if (argv[i] === "--password") password = takeValue(argv, i++, argv[i]);
    else {
      printError(`Unexpected argument: ${argv[i]}`);
      return 2;
    }
  }
  if (!handle || !name || !description) {
    printError("Usage: profile update HANDLE --name NAME --description TEXT [--password PASSWORD]");
    return 2;
  }
  if (!password) {
    printError("Missing password: use --password or set PDS_POST_PASSWORD.");
    return 2;
  }
  const session = await login(handle, password);
  if (!session) return 1;
  const result = await createRecord(session, "app.bsky.actor.profile", {
    "$type": "app.bsky.actor.profile",
    displayName: name,
    description,
  });
  if (!result) return 1;
  printSuccess("Profile updated!");
  return 0;
}

function help(): number {
  console.log(`Usage:
  scripts/dev/pds_cli.ts [--json] account create HANDLE EMAIL [--password PASSWORD]
  scripts/dev/pds_cli.ts [--json] post create HANDLE TEXT [--password PASSWORD]
  scripts/dev/pds_cli.ts [--json] profile update HANDLE --name NAME --description TEXT [--password PASSWORD]
`);
  return 2;
}

async function main(): Promise<number> {
  const argv = stripWrapperFlags(Deno.args);
  configureOutput();
  const [command, subcommand, ...rest] = argv;
  if (command === "account" && subcommand === "create") return await handleAccountCreate(rest);
  if (command === "post" && subcommand === "create") return await handlePostCreate(rest);
  if (command === "profile" && subcommand === "update") return await handleProfileUpdate(rest);
  return help();
}

if (import.meta.main) {
  Deno.exit(await main());
}
